pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/relocation/SegmentTierAssigner.java

```diff
@@ -0,0 +1,112 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.relocation;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.Collections;
+import java.util.List;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.pinot.common.metadata.segment.SegmentZKMetadata;
+import org.apache.pinot.common.metrics.ControllerMetrics;
+import org.apache.pinot.common.tier.Tier;
+import org.apache.pinot.common.tier.TierSegmentSelector;
+import org.apache.pinot.common.utils.config.TierConfigUtils;
+import org.apache.pinot.controller.ControllerConf;
+import org.apache.pinot.controller.LeadControllerManager;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.controller.helix.core.periodictask.ControllerPeriodicTask;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Periodic task to calculate the target tier the segment belongs to and set it into segment ZK metadata as goal
+ * state, which can be checked by servers when loading the segment to put it onto the target storage tier.
+ */
+public class SegmentTierAssigner extends ControllerPeriodicTask<Void> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentTierAssigner.class);
+
+  public SegmentTierAssigner(PinotHelixResourceManager pinotHelixResourceManager,
+      LeadControllerManager leadControllerManager, ControllerConf config, ControllerMetrics controllerMetrics) {
+    super(SegmentTierAssigner.class.getSimpleName(), config.getSegmentTierAssignerFrequencyInSeconds(),
+        config.getSegmentTierAssignerInitialDelayInSeconds(), pinotHelixResourceManager, leadControllerManager,
+        controllerMetrics);
+  }
+
+  @Override
+  protected void processTable(String tableNameWithType) {
+    TableConfig tableConfig = _pinotHelixResourceManager.getTableConfig(tableNameWithType);
+    Preconditions.checkState(tableConfig != null, "Failed to find table config for table: {}", tableNameWithType);
+    List<Tier> sortedTiers;
+    if (CollectionUtils.isEmpty(tableConfig.getTierConfigsList())) {
+      LOGGER.info("No tierConfigs so use default tier for segments of table: {}", tableNameWithType);
+      sortedTiers = Collections.emptyList();
+    } else {
+      LOGGER.info("Checking and updating target tiers for segments of table: {}", tableNameWithType);
+      sortedTiers = TierConfigUtils
+          .getSortedTiers(tableConfig.getTierConfigsList(), _pinotHelixResourceManager.getHelixZkManager());
+      LOGGER.debug("Sorted tiers: {} configured for table: {}", sortedTiers, tableNameWithType);
+    }
+    for (String segmentName : _pinotHelixResourceManager.getSegmentsFor(tableNameWithType, true)) {
```

Comment:
We should skip the replaced segments

Reply:
(minor) Batch read the segment ZK metadata instead of reading them one by one under the for loop

Reply:
the `true` flag is `shouldExcludeReplacedSegments`.